File: src/mantle/world/CoordTuple.scala
package mantle.world

import java.lang.String

/**
 * CoordTuple
 *
 * @param x The x coordinate
 * @param y The y coordinate
 * @param z The z coordinate
 *
 * @author mDiyo
 * @author Sunstrike <sunstrike@azurenode.net>
 */
class CoordTuple(val x: Int, val y:Int, val z:Int) extends Comparable[CoordTuple] {

  def this(posX: Double, posY: Double, posZ: Double) = this(Math.floor(posX).asInstanceOf[Int], Math.floor(posY).asInstanceOf[Int], Math.floor(posZ).asInstanceOf[Int])

  def this(tuple: CoordTuple) = this(tuple.x, tuple.y, tuple.z)

  def equalCoords(posX: Int, posY: Int, posZ: Int): Boolean = this.x == posX && this.y == posY && this.z == posZ

  override def hashCode: Int = {
    val prime: Int = 31
    var result: Int = 1
    result = prime * result + x
    result = prime * result + y
    result = prime * result + z

    result
  }

  override def toString: String = s"X: $x, Y: $y, Z: $z"

  def compareTo(o: CoordTuple): Int = o match {
    case null         => throw new NullPointerException("Object cannot be null")
    case _ if x < o.x => -1
    case _ if x > o.x => 1
    case _ if y < o.y => -1
    case _ if y > o.y => 1
    case _ if z < o.z => -1
    case _ if z > o.z => 1
    case _            => 0
  }

}

File: src/mantle/world/CoordTuple.java
package mantle.world;

import java.lang.String;

/**
 * CoordTuple
 *
 * @author mDiyo
 */
public class CoordTuple implements Comparable
{
    public final int x;
    public final int y;
    public final int z;

    public CoordTuple(double posX, double posY, double posZ)
    {
        x = (int) Math.floor(posX);
        y = (int) Math.floor(posY);
        z = (int) Math.floor(posZ);
    }

    public CoordTuple(CoordTuple tuple)
    {
        x = tuple.x;
        y = tuple.y;
        z = tuple.z;
    }

    public boolean equalCoords (int posX, int posY, int posZ)
    {
        if (this.x == posX && this.y == posY && this.z == posZ)
            return true;
        else
            return false;
    }

    @Override
    public int hashCode ()
    {
        final int prime = 31;
        int result = 1;
        result = prime * result + x;
        result = prime * result + y;
        result = prime * result + z;
        return result;
    }

    public String toString ()
    {
        return "X: " + x + ", Y: " + y + ", Z: " + z;
    }

    @Override
    public int compareTo (Object o)
    {
        if (o == null)
            throw new NullPointerException("Object cannot be null");

        CoordTuple coord = (CoordTuple) o;

        if (x < coord.x)
        {
            return -1;
        }
        if (x > coord.x)
        {
            return 1;
        }
        if (y < coord.y)
        {
            return -1;
        }
        if (y > coord.y)
        {
            return 1;
        }
        if (z < coord.z)
        {
            return -1;
        }
        if (z > coord.z)
        {
            return 1;
        }

        return 0;
    }

}
